File: backend/services/post-ktor-mongo-fdm/src/main/kotlin/me/jason5lee/post_ktor_mongo_fdm/get_user/impl.kt
package me.jason5lee.post_ktor_mongo_fdm.get_user

import me.jason5lee.post_ktor_mongo_fdm.common.Time
import me.jason5lee.post_ktor_mongo_fdm.common.UserName
import me.jason5lee.post_ktor_mongo_fdm.common.utils.Db
import me.jason5lee.post_ktor_mongo_fdm.common.utils.Deps

class WorkflowImpl(private val deps: Deps) : Workflow(), ErrorsImpl {
    override suspend fun run(input: Query): User {
        val doc = Db.findById(
            db = deps.mongoDb,
            collection = Db.users,
            id = Db.toObjectIdOrNull(input.value) ?: throw userNotFound(),
            projection = null,
        ) ?: throw userNotFound()
        return User(
            name = UserName(doc.get("name")),
            creationTime = Time(doc.get("creationTime"))
        )
    }
}

File: backend/services/post-ts-koa-mongo/src/get-user/impl.ts
import { UserId, UserName, Time } from "../common";
import { Deps } from "../common/utils";
import { Workflow } from ".";
import { WithId } from "mongodb";
import * as db from "../common/utils/db";
import { errors } from "./api";
import * as runtypes from "runtypes";

export class WorkflowImpl implements Workflow {
    private static readonly expectedUser = runtypes.Record({
        name: runtypes.String,
        creationTime: runtypes.Number,
    });
    async run(id: UserId): Promise<{ userName: UserName, creationTime: Time }> {
        const oid = db.tryParseId(id);
        if (oid === undefined) {
            throw errors.userNotFound();
        }
        const user: WithId<unknown> | null = await this.deps.mongoDb.collection(db.users).findOne({ _id: oid });
        if (user === null) {
            throw this.errors.userNotFound();
        }
        db.validate(WorkflowImpl.expectedUser, db.users, user);
        return {
            userName: user.name as UserName,
            creationTime: { utc: user.creationTime } as Time,
        };
    }

    readonly errors = errors;
    constructor(private readonly deps: Deps) { }
}
